### PR TITLE
feat: add Muon optimizer support

### PR DESCRIPTION
**Pull request by xxyux**
### PR Category

Execute Infrastructure 

### PR Types

New features

### Description

Add `paddle.optimizer.Muon`, a Newton-Schulz orthogonalisation-based optimizer
that applies orthogonal updates to 2-D & 3-D parameter matrices (e.g. Linear weights, MoE weights)
while automatically falling back to AdamW for non-2D parameters (embeddings,
biases, layer norms).

#### Core optimizer (`python/paddle/optimizer/muon.py`)
- Newton-Schulz iteration (5th-order, 6 steps by default) for computing the
  orthogonal projection of gradients
- Per-head QKV split: when a fused QKV weight is annotated with `needs_qkv_split`,
  each attention head is orthogonalised independently
- `sharded_state_dict` support for distributed checkpointing

#### ShardingV2 integration
- `muon_sharding_annotations.py`: annotate each 1-D parameter shard with
  original shape, sharding indices, and gather metadata
- `dygraph_sharding_optimizer.py`: call annotation helpers during parameter
  registration; sort Muon params for better GPU memory locality

#### ShardingV3 implementation (`dygraph_sharding_optimizer_v3.py`)
- Full-parameter ownership model: each 2-D Muon parameter is owned by one
  sharding rank (round-robin assignment)
- Forward: broadcast owned params to all ranks
- Backward: reduce gradients to owner rank, apply Muon update locally, then
  broadcast updated params
- Non-Muon parameters use conventional element-wise sharding with allgather

#### Shared utilities (`muon_comm_utils.py`)
- `should_use_muon(name, shape)`: determine whether a parameter should use
  Muon based on its name and dimensionality
- `gather_varlen`: variable-length allgather for V2 split-param sharding
- `get_sharding_info`: compute per-rank element counts and offsets

#### Tests
- `test_parallel_dygraph_muon.py` with 3 test cases:
  - `test_muon_sharding`: ShardingV2 + AMP O2 vs single-device reference
  - `test_muon_sharding_v3`: ShardingV3 + AMP O2 vs single-device reference
  - `test_muon_mp_qkv_split`: Tensor parallel + sharding with fused QKV
- All tests run on 2 GPUs and validate with `assert_allclose`

### 是否引起精度变化

否

### CODE REVIEW

File: python/paddle/distributed/fleet/utils/muon_comm_utils.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+def should_use_muon(name, shape):
+    """Return True if a parameter should receive Muon (orthogonal) updates.
+
+    Muon applies only to 2-D weight matrices.  Embeddings, biases, and
+    LM-head weights fall back to AdamW.
+    """
+    if len(shape) != 2:
+        return False
+    name = name.lower()
+    if "embed" in name or "bias" in name or "lm_head" in name:
```
Comment (GuoxiaWang) [inline python/paddle/distributed/fleet/utils/muon_comm_utils.py:147]: 这个地方需要非常小心，未来应该需要用户在外面也能指定 should_use_muon 的规则，不然这已经写死到框架里了，用户改不动。

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer.py [hunk 2]
```diff
@@ -1268,8 +1277,82 @@ def step(self):
...
                 if grad_var is not None:
+                    if is_muon:
+                        # Lazy import to avoid circular dependency.
+                        from ...utils.muon_comm_utils import (
+                            get_sharding_info,
+                            should_use_muon,
+                        )
+
+                        original_p = global_param_map[param.name]
+                        if should_use_muon(original_p.name, original_p.shape):
```
Comment (GuoxiaWang) [inline python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer.py]: 这一坨函数，好像跟上下没有耦合，是不是可以抽取到 muon_comm_utils 中呢？

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer.py [hunk 3]
```diff
@@ -1268,8 +1277,82 @@ def step(self):
...
+                                    sharding_world_size,
+                                    sharding_rank,
+                                )
+                                param.sharding_indices = indices
+                                param.sharding_my_offset = my_offset
+
                     params_grads.append((param, grad_var))
 
+            if is_muon and params_grads:
+                import numpy as np
```
Comment (GuoxiaWang) [inline python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer.py]: 这一个 numpy import 也需要 lazy import 吗？是否可以放到前面去

File: python/paddle/optimizer/muon.py [hunk 4]
```diff
@@ -0,0 +1,731 @@
...
+                )
+
+    def _create_accumulators(self, block, parameters):
+        """Standard entry-point used by checkpoint-resume infrastructure.
+
+        Delegates to _ensure_accumulators so that accumulators are keyed by
+        the slice_param name (consistent with _apply_optimize) rather than by
+        a master-weight name, which would cause a key mismatch under AMP O2.
+        """
+        from paddle.distributed.fleet.utils.muon_comm_utils import (
```
Comment (GuoxiaWang) [inline python/paddle/optimizer/muon.py]: 这里是不是导入多余了，可以去除，文件开头处已经 import 过了

**Review by GuoxiaWang — CHANGES_REQUESTED** [review]

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/hybrid_parallel_optimizer.py [hunk 5]
```diff
@@ -284,11 +287,13 @@ def __init__(self, optimizer, hcg, strategy):
             split_param = strategy.hybrid_configs[
                 'sharding_configs'
             ].split_param
-            ShardingOptimizer = (
-                DygraphShardingOptimizerV2
-                if split_param
-                else DygraphShardingOptimizer
-            )
+            use_sharding_v3 = os.environ.get("FLAGS_sharding_v3", "0") == "1"
```
Comment (GuoxiaWang) [inline python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/hybrid_parallel_optimizer.py:290]: 未来需要统一一个参数控制才行。

**Review by GuoxiaWang — APPROVED** [review]
LGTM，有2个修改后续 PR 解决

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer_v3.py [hunk 6]
```diff
@@ -0,0 +1,1068 @@
...
+    FusedCommBuffer,
+    assign_group_by_size,
+)
+
+
+def _is_trainable(param):
+    return not param.stop_gradient
+
+
+class DygraphShardingOptimizerV3:
```
Comment (sneaxiy) [inline python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer_v3.py:60]: Better rename it to be `MuonShardingOptimizer`.

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer_v3.py [hunk 7]
```diff
@@ -0,0 +1,1068 @@
...
+        for p in self._parameter_list:
+            if not _is_trainable(p):
+                continue
+            color = getattr(p, 'color', -1)
+            if isinstance(color, dict):
+                color_val = color.get('color', -1)
+            else:
+                color_val = color
+
+            if color_val == 'moe_expert':
```
Comment (sneaxiy) [inline python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer_v3.py:149]: Why does Muon optimizer need to distinguish the MoE expert parameters with others?

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer_v3.py [hunk 8]
```diff
@@ -0,0 +1,1068 @@
...
+            logger.info(
+                f"[V3-init rank={self._sharding_rank}]   1d_adamw: "
+                f"name={_p.name!r} shape={list(_p.shape)} "
+                f"should_muon={should_use_muon(_p.name, _p.shape)} "
+                f"color={getattr(_p, 'color', None)} "
+                f"comm_group_nranks={_cg_nranks}"
+            )
+
+        # --- [SLICE SIZE SUMMARY] Per-rank slice param sizes within this PP stage ---
+        import math as _math
```
Comment (sneaxiy) [inline python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer_v3.py:328]: Why do we need to `import math` inside this function instead of in the front of this file.

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/hybrid_parallel_optimizer.py (hunk 5 above)
Comment (sneaxiy) [inline python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/hybrid_parallel_optimizer.py:290]: Do not use `FLAGS_sharding_v3` environment variable. Should use `DistributedStrategy` instead.

File: python/paddle/distributed/fleet/utils/muon_comm_utils.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        length = max(0, end - start)
+        indices.append(length)
+        if r == rank:
+            my_slice_offset = current_relative_offset
+        current_relative_offset += length
+
+    return indices, my_slice_offset
+
+
+def should_use_muon(name, shape):
```
Comment (sneaxiy) [inline python/paddle/distributed/fleet/utils/muon_comm_utils.py:136]: Should be the argument of `MuonShardingOptimizer` and `optimizer.Muon`.

**Review by sneaxiy — COMMENTED** [review]
Moreover, should consider how to deal with `fused_qkv` and `fused_up_gate_proj` when tensor_parallel_degree > 1.
